java/cdk/src/main/java/sleeper/cdk/util/AutoStopEcsClusterTasks.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2022-2025 Crown Copyright
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package sleeper.cdk.util;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import software.amazon.awscdk.CustomResource;
+import software.amazon.awscdk.Duration;
+import software.amazon.awscdk.customresources.Provider;
+import software.amazon.awscdk.services.ecs.ICluster;
+import software.amazon.awscdk.services.iam.IRole;
+import software.amazon.awscdk.services.iam.ManagedPolicy;
+import software.amazon.awscdk.services.iam.PolicyStatement;
+import software.amazon.awscdk.services.lambda.IFunction;
+import software.amazon.awscdk.services.logs.ILogGroup;
+import software.constructs.Construct;
+
+import sleeper.cdk.jars.LambdaCode;
+import sleeper.core.deploy.LambdaHandler;
+import sleeper.core.properties.instance.InstanceProperties;
+import sleeper.core.util.EnvironmentUtils;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+@SuppressFBWarnings("NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE")
+public class AutoStopEcsClusterTasks {
+
+    private AutoStopEcsClusterTasks() {
+    }
+
+    public static void autoStopTasksOnEcsCluster(
+            Construct scope, InstanceProperties instanceProperties, LambdaCode lambdaCode,
+            ICluster cluster, String clusterName,
+            ILogGroup logGroup,
+            ILogGroup providerLogGroup) {
+
+        String id = cluster.getNode().getId() + "-AutoStop";
+        String functionName = clusterName + "-autostop";
+
+        IFunction lambda = lambdaCode.buildFunction(scope, LambdaHandler.AUTO_STOP_ECS_CLUSTER_TASKS, id + "Lambda", builder -> builder
```

Comment:
I wonder a little whether it might be worth deploying the lambdas and providers for the custom CDK resources in a separate stack. I'm sure what you've written is fine as it is. Do you think it might be worth refactoring that as a separate issue?